contracts/margin/external/exchangewrappers/ZeroExV2ExchangeWrapper.sol

```diff
@@ -0,0 +1,295 @@
+    )
+        private
+    {
+        uint256 takerFee = MathHelpers.getPartialAmount(
+            requestedFillAmount,
+            order.takerAssetAmount,
+            order.takerFee
+        );
+
+        if (takerFee == 0) {
+            return;
+        }
+
+        require(
+            TRUSTED_MSG_SENDER[msg.sender],
+            "ZeroExV2ExchangeWrapper#transferTakerFee: Only trusted senders can dictate the fee payer"
+        );
+
+        ZRX.transferFrom(
+            tradeOriginator,
+            address(this),
+            takerFee
+        );
+    }
+
+    function parseSignature(
+        bytes orderData
+    )
+        private
+        pure
+        returns (bytes memory)
+    {
+        bytes memory signature = new bytes(66);
+
+        /* solium-disable-next-line security/no-inline-assembly */
+        assembly {
+            mstore(add(signature, 32), mload(add(orderData, 352)))  // first 32 bytes
+            mstore(add(signature, 64), mload(add(orderData, 384)))  // next 32 bytes
+            mstore(add(signature, 66), mload(add(orderData, 386)))  // last 2 bytes
+        }
+
+        return signature;
+    }
+
+    function parseOrder(
+        bytes orderData,
+        address makerToken,
+        address takerToken
+    )
+        private
+        pure
+        returns (Order memory)
+    {
+        Order memory order;
+
+        /* solium-disable-next-line security/no-inline-assembly */
+        assembly {
+            mstore(order,           mload(add(orderData, 32)))  // maker
+            mstore(add(order, 32),  mload(add(orderData, 64)))  // taker
+            mstore(add(order, 64),  mload(add(orderData, 96)))  // feeRecipient
+            mstore(add(order, 96),  mload(add(orderData, 128))) // sender
+            mstore(add(order, 128), mload(add(orderData, 160))) // makerTokenAmount
+            mstore(add(order, 160), mload(add(orderData, 192))) // takerTokenAmount
+            mstore(add(order, 192), mload(add(orderData, 224))) // makerFee
+            mstore(add(order, 224), mload(add(orderData, 256))) // takerFee
+            mstore(add(order, 256), mload(add(orderData, 288))) // expirationUnixTimestampSec
+            mstore(add(order, 288), mload(add(orderData, 320))) // salt
+        }
+
+        order.makerAssetData = tokenAddressToAssetData(makerToken);
+        order.takerAssetData = tokenAddressToAssetData(takerToken);
+
+        return order;
+    }
+
+    function tokenAddressToAssetData(
+        address tokenAddress
+    )
+        private
+        pure
+        returns (bytes)
+    {
+        bytes memory result = new bytes(36);
+
+        // padded version of bytes4(keccak256("ERC20Token(address)"));
+        bytes32 selector = 0xf47261b000000000000000000000000000000000000000000000000000000000;
+
+        /* solium-disable-next-line security/no-inline-assembly */
+        assembly {
+            // Store the neighbors and address into memory
+            mstore(add(result, 32), selector)
```

Comment:
put a comment that you're starting 32 bytes after result to account for the first 32 bytes storing the array length. This confused me for a minute